devito/passes/clusters/blocking.py

```diff
@@ -44,62 +46,206 @@ def blocking(clusters, options):
+                return clusters
+
+            # Heuristic: same as above if there's a local SubDimension
+            if any(i.dim.is_Sub and i.dim.local for i in c.itintervals):
+                return clusters
+
+        if len(clusters) > 1:
+            # Heuristic: same as above if it induces dynamic bounds
+            exprs = flatten(c.exprs for c in as_tuple(clusters))
+            scope = Scope(exprs)
+            if any(i.is_lex_non_stmt for i in scope.d_all_gen()):
+                return clusters
+        else:
+            # Just avoiding potentially expensive checks
+            pass
+
+        # All good, `d` is actually TILABLE
+        processed = attach_property(clusters, d, TILABLE)
+
+        return processed
+
+
+class AnalyzeSkewing(Queue):
+
+    """
+    Encode the SKEWABLE Dimensions.
+    """
+
+    def callback(self, clusters, prefix):
+        if not prefix:
+            return clusters
+
+        d = prefix[-1].dim
+
+        for c in clusters:
+            if TILABLE not in c.properties[d]:
+                return clusters
+
+        processed = attach_property(clusters, d, SKEWABLE)
+
+        return processed
+
+
+class SynthesizeBlocking(Queue):
+
+    template = "%s%d_blk%s"
+
+    def __init__(self, sregistry, options):
+        self.sregistry = sregistry
+
         self.levels = options['blocklevels']
 
-        self.nblocked = Counter()
+        # A tool to unroll the explicit integer block shapes, should there be any
+        if options['par-tile']:
+            self.generator = UnboundedMultiTuple(*options['par-tile'])
+        else:
+            self.generator = None
+
+        super().__init__()
 
-        super(Blocking, self).__init__()
+    def process(self, clusters):
+        return self._process_fatd(clusters, 1)
 
     def _make_key_hook(self, cluster, level):
         return (tuple(cluster.guards.get(i.dim) for i in cluster.itintervals[:level]),)
 
-    def _process_fdta(self, clusters, level, prefix=None):
-        # Truncate recursion in case of TILABLE, non-perfect sub-nests, as
-        # it's an unsupported case
-        if prefix:
-            d = prefix[-1].dim
-            test0 = any(TILABLE in c.properties[d] for c in clusters)
-            test1 = len({c.itintervals[:level] for c in clusters}) > 1
-            if test0 and test1:
-                return self.callback(clusters, prefix)
-
-        return super(Blocking, self)._process_fdta(clusters, level, prefix)
-
     def callback(self, clusters, prefix):
         if not prefix:
             return clusters
 
         d = prefix[-1].dim
 
+        if not any(TILABLE in c.properties[d] for c in clusters):
+            return clusters
+
         # Create the block Dimensions (in total `self.levels` Dimensions)
-        name = self.template % (d.name, self.nblocked[d], '%d')
+        base = self.sregistry.make_name(prefix=d.name)
+
+        if self.generator:
+            # An explicit integer step has been supplied
```

Comment:
Something is missing here? Sounds weird?

Reply:
Fixing